File: leetcode-go/Array/LC0026-删除有序数组中的重复项.go
package main

import "fmt"

/**
 * Question: LC26. 删除有序数组中的重复项
 *
 * Input   : nums = [0,0,1,1,1,2,2,3,3,4]
 * Output  : 5, nums = [0,1,2,3,4,_,_,_,_,_]
 * Explain : 不要使用额外的空间，你必须在 原地 修改输入数组 并在使用 O(1) 额外空间的条件下完成。
 *
 * @Date: 2022-06-14
 * @Author: CoulsonZero
 */

// 快慢指针法
func removeDuplicates(nums []int) int {
	l := 0
	for r := 1; r < len(nums); r++ {
		if nums[l] != nums[r] {
			l++
			nums[l] = nums[r]
		}
	}
	return l + 1
}

func main() {
	nums := []int{0, 0, 1, 1, 1, 2, 2, 3, 3, 4}
	fmt.Println(removeDuplicates(nums)) // len: 5
}

File: leetcode-python/Array/LC0026-删除有序数组中的重复项.py
'''
lc26. 删除排序数组中的重复项
input: nums = [0,0,1,1,1,2,2,3,3,4]
output: 5, nums = [0,1,2,3,4]

标签：数组 双指针 从后遍历删除
'''
class Solution:
    @staticmethod
    def removeDuplicates(nums):
        for i in range(len(nums)-1, -1, -1):
            if nums[i] == nums[i-1]:
                del nums[i]
        return len(nums)

class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        l = 1
        for r in range(1, len(nums)):
            if nums[r] != nums[r-1]:
                nums[l] = nums[r]
                l += 1
        return l


if __name__ == '__main__':
    nums = [0, 0, 1, 1, 1, 2, 2, 3, 3, 4]
    print(Solution.removeDuplicates(nums))

